### ISSUE TITLE
Terrarium Hardware Configuration

### ISSUE THREAD

**Issue by DaveMulls** [bkshepherd/DaisySeedProjects#40]
Hey there, I'm trying to get this working on a Pedalpcb Terrarium build, I've been speaking to @xconverge who has been kind enough to get me to the point where I have the project loaded on the seed and I can pass clean signal through, but I have no effects and no response from footswitches, LED's, and toggle switches. 

I'm not particularly worried about switching between modules as I am happy to play with one at a time, but I was hoping to sort out the controls mapping. @xconverge suggested I tag @bkshepherd or @keyth72 to see if you guys had any ideas. Thanks in advance!

**Comment by xconverge** [comment]
@bkshepherd @keyth72 I am totally onboard to support @DaveMulls with getting the one effect he is interested in working, etc, but I am a bit at a loss for why the tremolo isnt working out of the box once he setup the terrarium header in guitar_pedal.cpp. I can't think of much and was wondering if either of you had a) terrarium hw you have tested recently with this project or b) any other guesses/suggestions

**Comment by DaveMulls** [comment]
Just to confirm, I have made these changes in guitar_pedal.cpp: 

#include "Hardware-Modules/guitar_pedal_terrarium.h"

// Hardware Interface
GuitarPedalTerrarium hardware;

**Comment by xconverge** [comment]
@DaveMulls the https://github.com/bkshepherd/DaisySeedProjects/blob/main/Software/GuitarPedal/Hardware-Modules/guitar_pedal_terrarium.cpp file doesn't look right to me at all for at least the current rev of the build docs I see for the terrarium

Can you try these changes:

```diff
diff --git a/Software/GuitarPedal/Hardware-Modules/guitar_pedal_terrarium.cpp b/Software/GuitarPedal/Hardware-Modules/guitar_pedal_terrarium.cpp
index 58ab1a3..ce63e0a 100644
--- a/Software/GuitarPedal/Hardware-Modules/guitar_pedal_terrarium.cpp
+++ b/Software/GuitarPedal/Hardware-Modules/guitar_pedal_terrarium.cpp
@@ -17,12 +17,12 @@ GuitarPedalTerrarium::~GuitarPedalTerrarium() {}
 void GuitarPedalTerrarium::Init(bool boost) {
     BaseHardwareModule::Init(boost);
 
-    Pin knobPins[] = {seed::D16, seed::D17, seed::D18, seed::D19, seed::D20, seed::D21};
+    Pin knobPins[] = {seed::D17, seed::D18, seed::D19, seed::D20, seed::D21, seed::D22};
     InitKnobs(6, knobPins);
 
-    Pin switchPins[] = {seed::D26, seed::D25, seed::D10, seed::D9, seed::D8, seed::D7};
-    InitSwitches(6, switchPins);
+    Pin switchPins[] = {seed::D27, seed::D26};
+    InitSwitches(2, switchPins);
 
-    Pin ledPins[] = {seed::D23, seed::D22};
+    Pin ledPins[] = {seed::D24, seed::D23};
     InitLeds(2, ledPins);
 }

```

**Comment by DaveMulls** [comment]
Hey @xconverge unfortunately no change with the alterations, still passing clean signal though. Here is my guitar_pedal_terrarium.cpp with the alterations, let me know if I maybe input them wrong: 

#include "guitar_pedal_terrarium.h"

using namespace bkshepherd;

static const int s_switchParamCount = 2;
static const PreferredSwitchMetaData s_switchMetaData[s_switchParamCount] = {
    {sfType : SpecialFunctionType::Bypass, switchMapping : 0}, {sfType : SpecialFunctionType::Alternate, switchMapping : 1}};

GuitarPedalTerrarium::GuitarPedalTerrarium() : BaseHardwareModule() {
    // Setup the Switch Meta Data for this hardware
    m_switchMetaDataParamCount = s_switchParamCount;
    m_switchMetaData = s_switchMetaData;
}

GuitarPedalTerrarium::~GuitarPedalTerrarium() {}

void GuitarPedalTerrarium::Init(bool boost) {
    BaseHardwareModule::Init(boost);

    Pin knobPins[] = {seed::D17, seed::D18, seed::D19, seed::D20, seed::D21, seed::D22};
    InitKnobs(6, knobPins);

    Pin switchPins[] = {seed::D27, seed::D26, seed::D11, seed::D10, seed::D9, seed::D8};
    InitSwitches(6, switchPins);

    Pin ledPins[] = {seed::D24, seed::D23};
    InitLeds(2, ledPins);
}


Are any changes required to the guitar_pedal_terrarium.h?

**Comment by xconverge** [comment]
Looking at his project here they have the same mappings https://github.com/schult/terrarium-poly-octave/tree/main so maybe I read the docs (https://docs.pedalpcb.com/project/Terrarium.pdf) wrong and you should just revert all of the changes I suggested in that file and ignore that idea of mine...

**Comment by bkshepherd** [comment]
Hmmm, it's been a long time since I tried this on the terrarium hardware, but it was working at the time.  I'll see if I can dig mine up over the holidays.   @xconverge Do you think some of the fancier foot switch input configurations stuff maybe changed the way something worked?

**Comment by xconverge** [comment]
> Hmmm, it's been a long time since I tried this on the terrarium hardware, but it was working at the time. I'll see if I can dig mine up over the holidays. @xconverge Do you think some of the fancier foot switch input configurations stuff maybe changed the way something worked?

I have looked through the code pretty thoroughly and can't find anything obvious, let me flash a "terrarium-like" config to mine and see if anything shows up

**Comment by bkshepherd** [comment]
The only other thing I can think of is it's not obeying the fact that Terrarium doesn't have the relayed based bypass.  The code for the Terrarium hardware module is setting m_supportsTrueBypass to false (which is correct), and m_supportsStereo to false (which is also correct) but maybe something in guitar_pedal.cpp is no longer working correctly when try bypass and stereo are off

**Comment by xconverge** [comment]
That is going to be it, I can kind of reproduce the problem with this , so should be easy for me to get it up and going

```diff
diff --git a/Software/GuitarPedal/Hardware-Modules/guitar_pedal_125b.cpp b/Software/GuitarPedal/Hardware-Modules/guitar_pedal_125b.cpp
index 95cc118..cef151c 100644
--- a/Software/GuitarPedal/Hardware-Modules/guitar_pedal_125b.cpp
+++ b/Software/GuitarPedal/Hardware-Modules/guitar_pedal_125b.cpp
@@ -31,7 +31,7 @@ void GuitarPedal125B::Init(bool boost) {
     Pin ledPins[] = {seed::D22, seed::D23};
     InitLeds(2, ledPins);
 
-    InitMidi(seed::D30, seed::D29);
-    InitDisplay(seed::D9, seed::D11);
-    InitTrueBypass(seed::D1, seed::D12);
+    // InitMidi(seed::D30, seed::D29);
+    // InitDisplay(seed::D9, seed::D11);
+    // InitTrueBypass(seed::D1, seed::D12);
 }
\ No newline at end of file

```

@DaveMulls do your LEDs work though? Because I can get the LEDs to work even though the signal is always clean

**Comment by bkshepherd** [comment]
@xconverge I'm not surprised that not initializing TrueBypass makes the 125B pass a clear signal though,  because if the relays aren't turning on, they default to bypass mode.  So that doesn't necessarily mean it's the same issue on the terrarium hardware, because by default on the terrarium all signal goes through the daisy signal path (there is no bypass at all)

**Comment by xconverge** [comment]
@DaveMulls you could try an earlier version of the code (and make the same tweak to guitar_pedal.cpp to use the terrarium config)

`git checkout v1.0.0` in the root of the repo to pull a version from a year ago

then make sure your submodules are pulled and do a `make clean && make -j8` in both libDaisy and daisySP

This is certainly something you can try... even if moving between submodule versions can get annoying with the rebuilds and cleanup

**Comment by DaveMulls** [comment]
@xconverge no lights, I'll give the earlier version a go and report back shortly! Thanks guys

**Comment by DaveMulls** [comment]
I know this is probably a mistake on my end as I don't have much experience with this stuff, I've reverted to v1.0.0 and successfully ran "make clean && make -j8" in both libDaisy and daisySP, and changed the hardware config for Terrarium, but when I try to flash to the seed I get this error: 

$ make
arm-none-eabi-g++ -c -mcpu=cortex-m7 -mthumb -mfpu=fpv5-d16 -mfloat-abi=hard -DUSE_HAL_DRIVER -DSTM32H750xx -DHSE_VALUE=16000000  -DCORE_CM7 -DSTM32H750IB -DARM_MATH_CM7 -DUSE_FULL_LL_DRIVER -include stm32h7xx.h -IlibDaisy -IlibDaisy/src/ -IlibDaisy/src/sys -IlibDaisy/src/usbd -IlibDaisy/src/usbh -IlibDaisy/Drivers/CMSIS_5/CMSIS/Core/Include/ -IlibDaisy/Drivers/CMSIS-DSP/Include -IlibDaisy/Drivers/CMSIS-Device/ST/STM32H7xx/Include -IlibDaisy/Drivers/STM32H7xx_HAL_Driver/Inc/ -IlibDaisy/Middlewares/ST/STM32_USB_Device_Library/Core/Inc -IlibDaisy/Middlewares/ST/STM32_USB_Host_Library/Core/Inc -IlibDaisy/Middlewares/ST/STM32_USB_Host_Library/Class/MSC/Inc -IlibDaisy/Middlewares/ST/STM32_USB_Host_Library/Class/MIDI/Inc -IlibDaisy/core/ -IDaisySP/Source -IlibDaisy/Middlewares/Third_Party/FatFs/src -O1 -Wall -Wno-missing-attributes -fasm -fdata-sections -ffunction-sections -Wno-stringop-overflow -g -ggdb -MMD -MP -MF"build/guitar_pedal.d" -fno-exceptions -fasm -finline -finline-functions-called-once -fshort-enums -fno-move-loop-invariants -fno-unwind-tables -fno-rtti -Wno-register -std=gnu++14 -Wa,-a,-ad,-alms=build/guitar_pedal.lst guitar_pedal.cpp -o build/guitar_pedal.o
In file included from guitar_pedal.cpp:10:
Effect-Modules/reverb_module.h:29:5: error: 'ReverbSc' does not name a type
   29 |     ReverbSc m_reverbStereo;
      |     ^~~~~~~~
make: *** [libDaisy/core/Makefile:286: build/guitar_pedal.o] Error 1

Any idea what's gone wrong?

**Comment by xconverge** [comment]
You may want to run make clean in the root directory as well, but that error is making it look like DaisySP is not the right version or something potentially

This is a little crazy but it seems to build fine for me

```
cd libDaisy/ && make clean && make -j8 && cd ../DaisySP && make clean && make -j8 &
& cd ../ && make clean && make -j8
```

if you do a `git status` you should only see your terrarium change (like 2 lines, the header include and the variable definition). If you see anything else your repository and/or submodules is in a weird state

**Comment by DaveMulls** [comment]
Ok I got it to flash, sadly still no functionality. One difference though, the USER light on the seed is flashing intermittently when it's powered up. I'm like 99% sure that wasn't happening with the updated build.

**Comment by DaveMulls** [comment]
@bkshepherd ah ya that's probably it. Yes hardware is for sure working, I'm been running projects for about a year on it without issue, though it's only been through the daisy web programmer with .bin files. 

No rush at all, I really appreciate you taking a look!

**Comment by xconverge** [comment]
Can you share the output of your `make program-dfu` command? Is it possible it isn't flashing the way you would expect? I am guessing that you could also take the .bin created and use the web programmer as you were already doing

**Comment by DaveMulls** [comment]
This is for the current revision: 

$ make program-dfu
dfu-util -a 0 -s 0x90040000:leave -D build/guitarpedal.bin -d ,0483:df11
dfu-util 0.10

Copyright 2005-2009 Weston Schmidt, Harald Welte and OpenMoko Inc.
Copyright 2010-2020 Tormod Volden and Stefan Schmidt
This program is Free Software and has ABSOLUTELY NO WARRANTY
Please report bugs to http://sourceforge.net/p/dfu-util/tickets/

Warning: Invalid DFU suffix signature
A valid DFU suffix will be required in a future dfu-util release!!!
Opening DFU capable USB device...
ID 0483:df11
Run-time device DFU version 011a
Claiming USB DFU Interface...
Setting Alternate Setting #0 ...
Determining device status: state = dfuIDLE, status = 0
dfuIDLE, continuing
DFU mode device DFU version 011a
Device returned transfer size 4096
DfuSe interface name: "Flash "
Downloading element to address = 0x90040000, size = 184580
Erase           [=========================] 100%       184580 bytes
Erase    done.
Download        [=========================] 100%       184580 bytes
Download done.
File downloaded successfully
Transitioning to dfuMANIFEST state

I also tried using the Daisy web programmer with the bin file but it's the same result, successful download but no controls.

**Comment by DaveMulls** [comment]
SUCCESS!!! I have the modulated tremolo up and running, the control mapping is a bit off but I can definitely sort that out. I have been fiddling with chatGPT and it suggested modifying guitar_pedal.cpp to add this to the code and it seemed to sort everything out: 
![image](https://github.com/user-attachments/assets/74a55814-3c63-48fa-aa64-ebdd31f96537)

I think I should be good to go from here. I really appreciate both of you taking time out of your lives to help me out!

**Comment by xconverge** [comment]
What is the change there?

Now you can switch back to the latest and try again :)

I am not sure why the controls would be off, they look right in the codebase (my previous comments were wrong I think)

**Comment by DaveMulls** [comment]
I have basically zero programming experience but what it flagged was that the effects modules weren't being initialized and that I had to move the initialization to be closer to the beginning of the code, which I guess makes it run earlier on in the process. Once I made that change everything fell into place. As for controls that was my mistake, they are all correct!

**Comment by GuitarML** [comment]
For what it's worth I was able to run the main branch on my terrarium this afternoon without modifying the code, the tremolo effect came up with the expected knob mapping and led action. Glad you got it running though!

**Comment by xconverge** [comment]
What operating system are you on? @DaveMulls 

Let me know when/if you go to run the latest code. I will run some tools (https://en.wikipedia.org/wiki/Static_program_analysis) over the code later to see if they catch anything else, I still can't see what is wrong or changed super clearly but I am also not questioning what is working!

**Comment by DaveMulls** [comment]
@GuitarML Ya I think it's very likely that the problem was on my end and i just took enough tries and fiddling to fix whatever I had gotten scrambled. The difference with what worked is that I added in the "void initialize effects" line of code in and moved int section above it. When I look at the code on the git repo it looks like it already has the "void initialize effects" so I think I must have somehow altered it previously. I've got the latest code to run now, I think it's probably not worth investigating if it's running on your side, but let me know if you'd like to dig in and I can provide anything needed!
